internal/store/postgres/activity_repository.go

```diff
@@ -0,0 +1,148 @@
+		db = db.Where(`"timestamp" <= ?`, *filter.TimestampLte)
+	}
+	if err := db.Find(&activities).Error; err != nil {
+		return nil, err
+	}
+
+	var results []*domain.Activity
+	for _, activity := range activities {
+		a := &domain.Activity{}
+		if err := activity.ToDomain(a); err != nil {
+			return nil, fmt.Errorf("failed to convert model %q to domain: %w", activity.ID, err)
+		}
+		results = append(results, a)
+	}
+	return results, nil
+}
+
+func (r *ActivityRepository) GetOne(ctx context.Context, id string) (*domain.Activity, error) {
+	var m model.Activity
+	if err := r.db.
+		WithContext(ctx).
+		Joins("Provider").
+		Joins("Resource").
+		Where(`"activities"."id" = ?`, id).
+		First(&m).Error; err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return nil, activity.ErrNotFound
+		}
+		return nil, err
+	}
+
+	a := &domain.Activity{}
+	if err := m.ToDomain(a); err != nil {
+		return nil, err
+	}
+
+	return a, nil
+}
+
+func (r *ActivityRepository) BulkUpsert(ctx context.Context, activities []*domain.Activity) error {
+	if len(activities) == 0 {
+		return nil
+	}
+
+	activityModels := make([]*model.Activity, len(activities))
+	uniqueResourcesMap := map[string]*model.Resource{}
+
+	for i, a := range activities {
+		activityModels[i] = &model.Activity{}
+		if err := activityModels[i].FromDomain(a); err != nil {
+			return fmt.Errorf("failed to convert domain to model: %w", err)
+		}
+
+		// use single resource reference for activities with same resource
+		if r := activityModels[i].Resource; r != nil {
+			key := getResourceUniqueURN(*r)
+			if _, exists := uniqueResourcesMap[key]; !exists {
+				uniqueResourcesMap[key] = r
+			} else {
+				activityModels[i].Resource = uniqueResourcesMap[key]
+			}
+		}
+	}
+
+	return r.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
+		// upsert resources separately to avoid resource upsertion duplicate issue
+		var resources []*model.Resource
+		for _, r := range uniqueResourcesMap {
+			resources = append(resources, r)
+		}
+		if len(resources) > 0 {
+			if err := tx.Create(resources).Error; err != nil {
+				return fmt.Errorf("failed to upsert resources: %w", err)
+			}
+		}
+		for _, a := range activityModels {
+			// set resource id after upsertion
+			if a.Resource != nil {
+				a.ResourceID = a.Resource.ID
+			}
+		}
+
+		if err := tx.Omit("Resource").
+			Clauses(clause.OnConflict{
+				Columns: []clause.Column{
+					{Name: "provider_id"},
+					{Name: "provider_activity_id"},
+				},
+				UpdateAll: true,
+			}).
+			Create(activityModels).Error; err != nil {
+			return fmt.Errorf("failed to upsert provider activities: %w", err)
+		}
+
+		for i, m := range activityModels {
+			if err := m.ToDomain(activities[i]); err != nil {
+				return fmt.Errorf("failed to convert model %q to domain: %w", m.ID, err)
+			}
+		}
+
+		return nil
+	})
+}
+
+func getResourceUniqueURN(r model.Resource) string {
+	return fmt.Sprintf(`%s/%s/%s/%s`, r.ProviderType, r.ProviderURN, r.Type, r.URN)
+}
```

Comment:
was also thinking the same way, but didn't go with that because this helper is only being used within this Upsert method, having it as the model's method will make it available to other places which using the model and was trying to avoid that only